### PR TITLE
Fix `injectRoute`

### LINKED ISSUE #7561: Injected route error in pnpm workspace packages more than one directory deep
### What version of `astro` are you using?

2.7.0, 2.7.1, 2.7.2, 2.7.3

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

pnpm

### What operating system are you using?

macOS / Stackblitz

### What browser are you using?

Firefox/Chrome/Safari

### Describe the Bug

In versions of Astro 2.7.0, 2.7.1, 2.7.2 and 2.7.3 (latest), we’re seeing the [starter template](https://github.com/withastro/starlight/tree/main/examples/basics) failing to build inside the Starlight monorepo (usage via `create astro` works fine as does `astro dev`). This seems to be a wider bug with pnpm workspaces.

Minimal reproduction:

1. Load the [StackBlitz repro](https://stackblitz.com/edit/github-kwocty-hrptj2)
2. It will install dependencies for you
3. `cd sites/docs` to move to the package using Starlight
4. `pnpm build` to try building the site

This will trigger an error during the generating static routes phase:

> Cannot find module `'/home/projects/github-kwocty-hrptj2/sites/renderers.mjs'` imported from `/home/projects/github-kwocty-hrptj2/sites/docs/dist/node_modules/.pnpm/@astrojs+starlight@0.4.1_astro@2.7.0/node_modules/@astrojs/starlight/404.astro.mjs`

For context: `404.astro.mjs` is being generated from Starlight’s [injected 404 route](https://github.com/withastro/starlight/blob/eec46b40f4d35fdfb63c1ae2f2b148c83a448faf/packages/starlight/index.ts#L38-L45).

If you move the site up a level to `docs/` (instead of `sites/docs/`) and update `pnpm-workspace.yaml` accordingly, the project builds fine, so seems to be related to folder depth? Or perhaps something accidentally works for packages one level deep but not for more nested packages?

Also reported here: https://github.com/withastro/starlight/issues/305

### What's the expected result?

`astro build` is able to successfully build the site.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-kwocty-hrptj2

### Participation

- [ ] I am willing to submit a pull request for this issue.

### LINKED ISSUE #7563: Injected scripts not injected to non-local injected routes
### What version of `astro` are you using?

2.7.3

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

pnpm

### What operating system are you using?

macOS / Stackblitz

### What browser are you using?

Firefox/Chrome/Safari

### Describe the Bug

I reported `injectScript` not to be working for injected routes in https://github.com/withastro/astro/issues/7203 which was fixed by https://github.com/withastro/astro/pull/7262.

However, the fix only works for routes that are local to a project because it determines if something is an injected route by trying to match the _resolved_ file paths of routes to the endpoint specified in user config. This works fine for local routes where `injectRoute({ entryPoint: 'src/my-route.astro' })` and the file path of `'src/my-route.astro'` will match, but fails for situations where the resolved path and the authored path don’t match 1:1, for example if the route comes from a third-party package like Starlight: `injectRoute({ entryPoint: '@astrojs/starlight/index.astro' })`.

This may also impact other cases where authored and on-disk file paths can differ, e.g. for Windows users.

For the Starlight case, @hippotastic tried adding some debug logging and could see that `@astrojs/starlight/index.astro` was trying to match against a path of `file:///D:/Dev/starlight/packages/starlight/index.astro` and therefore failing to match.

Maybe a more fundamental fix in `createRouteManifest` is needed to specify injected routes early and not try to guess based on file path later?

The reproduction demonstrates this bug with a monorepo with one package that provides the route and one package that imports and renders the route on a page — where injected styles/scripts load successfully — and injects the route, where injected resources don’t load.

#### Reproduction instructions

```sh
# install deps
pnpm i
# move to the Astro site package
cd packages/site
# start the dev server
pnpm dev
```

The initial page shows Tailwind styles and an `alert()` added using `injectScript()`.

Navigate to the “Injected route” link to see the same page, which will be missing the injected Tailwind styles and the `alert()` call.

### What's the expected result?

Scripts or styles added with `injectScript` work on injected routes.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-ssvscu-zyglof?file=README.md

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Adds an additional `resolvedI jectedRoutes` array to the internal `Settings` instance. 
- Updates the Vite plugin that manages Astro Integrations to ensure `resolvedInjectedRoutes` is populated with a fully resolved file URL for each entry point.
- Fixes #7563 
- Attempted to address #7561 but that’s going back to the drawing board.

## Testing

Manually confirmed, hard to test because it’s a workspace/dependency thing. 
 
## Docs

Bug fix only

### CODE REVIEW

**Review by matthewp — APPROVED** [review]

File: .changeset/stupid-pants-press.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': patch
+---
+
+Ensure that `injectRoute` is properly detected in monorepos and from node_modules
```
Comment (delucis) [inline .changeset/stupid-pants-press.md:5]: I guess this changeset would be more accurate now:

```suggestion
Ensure that injected routes from `node_modules` are properly detected
```

File: packages/astro/src/vite-plugin-scripts/page-ssr.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ export default function astroScriptsPostPlugin({
 	return {
 		name: 'astro:scripts:page-ssr',
 		enforce: 'post',
-		transform(this, code, id, options) {
+		async transform(this, code, id, options) {
```
Comment (delucis) [inline packages/astro/src/vite-plugin-scripts/page-ssr.ts:15]: Nit, but I guess this is no longer doing anything `async`?
```suggestion
		transform(this, code, id, options) {
```

**Review by delucis — APPROVED** [review]
Thanks again for tackling @natemoo-re! Left two nits, but LGTM!

File: packages/astro/src/vite-plugin-scripts/page-ssr.ts (hunk 2 above)
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-scripts/page-ssr.ts:15]: Good catch!

File: packages/astro/src/vite-plugin-integrations-container/index.ts [hunk 3]
```diff
@@ -11,10 +11,28 @@ export default function astroIntegrationsContainerPlugin({
...
 }): VitePlugin {
+	let hasResolvedInjectedRoutes = false;
 	return {
 		name: 'astro:integration-container',
 		configureServer(server) {
 			runHookServerSetup({ config: settings.config, server, logging });
 		},
+		// Run `load` hook once just to get access to `PluginContext.resolve`
+		// which allows us to resolve each injectedRoute entryPoint to a real file URL on disk
+		async load() {
```
Comment (bluwy) [inline packages/astro/src/vite-plugin-integrations-container/index.ts:22]: I think we can also call this in `buildStart()`?

**Review by bluwy — APPROVED** [review]
One small nit, else this lgtm!

File: packages/astro/src/vite-plugin-integrations-container/index.ts (hunk 3 above)
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-integrations-container/index.ts]: Great call, thanks!

**Pull request merged**